src/Pos/ExecMode/Context.hs

```diff
@@ -0,0 +1,120 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell     #-}
+{-# LANGUAGE TypeOperators       #-}
+
+module Pos.ExecMode.Context
+    ( (:::)
+    , HasLens(..)
+    , modeContext
+    ) where
+
+import           Universum
+
+import           Language.Haskell.TH
+
+import           Pos.Util.Util       (HasLens (..), lensOf')
+
+-- | Bundle a tag and a value. The 'modeContext' function will transform
+-- @Tag ::: Integer@ to @Integer@ and use @Tag@ as meta-info about the field.
+data tag ::: value
+
+{- | Generate a mode context record with 'HasLens' instances.
+
+The 'modeContext' function takes a declaration like this:
+
+@
+    data PatakCtx ssc = PatakCtx
+        !(Buba    ::: Buba)
+        !(HuhaTag ::: Huha ssc)
+        !Bardaq
```

Comment:
To add a field to the context, use `:::`, it's like record fields, but tags instead of field names.

If you want to extend an existing context (like `RealModeContext` extends `NodeContext`), don't use `(:::)`. The limitation is that you can use it at most once.